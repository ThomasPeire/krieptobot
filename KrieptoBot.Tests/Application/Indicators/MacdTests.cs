using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using AwesomeAssertions;
using KrieptoBot.Application.Indicators;
using KrieptoBot.DataVisualizer;
using KrieptoBot.DataVisualizer.Extensions;
using KrieptoBot.Domain.Trading.ValueObjects;
using KrieptoBot.Infrastructure.Bitvavo.Dtos;
using KrieptoBot.Infrastructure.Bitvavo.Extensions;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NUnit.Framework;
using Plotly.NET;
using Snapshooter.NUnit;
using Color = System.Drawing.Color;

namespace KrieptoBot.Tests.Application.Indicators;

internal class MacdTests
{
    private IEnumerable<Candle> _candles;

    [SetUp]
    public void Setup()
    {
        InitCandles();
    }

    private void InitCandles()
    {
        var candlesJson = File.ReadAllText(@"./MockData/Bitvavo/candles_btc-eur.json");
        var deserializedCandles = JsonConvert.DeserializeObject(candlesJson) as JArray;
        _candles = deserializedCandles.Select(x =>
            new CandleDto
            {
                TimeStamp = x.Value<long>(0),
                Open = x.Value<decimal>(1),
                High = x.Value<decimal>(2),
                Low = x.Value<decimal>(3),
                Close = x.Value<decimal>(4),
                Volume = x.Value<decimal>(5)
            }.ConvertToKrieptoBotModel()).DistinctBy(x => x.TimeStamp);
    }

    [Test]
    public void Macd_Should_CalculateMacdValues()
    {
        var datetimeFrom = new DateTime(2021, 11, 01);
        var dateTimeTo = datetimeFrom.AddHours(8);

        var candlesToWorkWith = _candles
            .Where(x => x.TimeStamp >= datetimeFrom && x.TimeStamp <= dateTimeTo)
            .OrderBy(x => x.TimeStamp).ToList();

        var result = new Macd(new ExponentialMovingAverage()).Calculate(candlesToWorkWith);

        var macdLine = candlesToWorkWith.ToDictionary(x => x.TimeStamp,
            x => result.MacdLine.TryGetValue(x.TimeStamp, out var value) ? value : 0);
        var signalLine = candlesToWorkWith.ToDictionary(x => x.TimeStamp,
            x => result.SignalLine.TryGetValue(x.TimeStamp, out var value) ? value : 0);
        var histogram = candlesToWorkWith.ToDictionary(x => x.TimeStamp,
            x => result.Histogram.TryGetValue(x.TimeStamp, out var value) ? value : 0);

        Snapshot.Match(result);

#if DEBUG
        var candleVisualizer = new CandlesVisualizer();
        var candleChart = candleVisualizer.Visualize(candlesToWorkWith);

        candleChart = candleChart.AddSubChartLine(macdLine).AddLineChart(signalLine, Color.Chartreuse).AddLineChart(histogram, Color.Gold);
        candleChart.WithSize(1920, 1080).WithConfig(Config.init(Responsive: true)).Show();
#endif
    }
}