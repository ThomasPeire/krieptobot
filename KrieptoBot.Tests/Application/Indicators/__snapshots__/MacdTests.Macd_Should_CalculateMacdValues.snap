{
  "MacdLine": {
    "2021-11-01T00:00:00Z": 0.0,
    "2021-11-01T00:05:00Z": 10.051282051282051282051281,
    "2021-11-01T00:10:00Z": 29.378625173496968368763240,
    "2021-11-01T00:15:00Z": 50.727641862713818298594395,
    "2021-11-01T00:20:00Z": 55.867439646428286838506618,
    "2021-11-01T00:25:00Z": 59.816085321458958351674239,
    "2021-11-01T00:30:00Z": 60.792191359486305453424662,
    "2021-11-01T00:35:00Z": 67.724557157262873063588958,
    "2021-11-01T00:40:00Z": 64.406900079654624062280768,
    "2021-11-01T00:45:00Z": 61.153386413954091379634399,
    "2021-11-01T00:50:00Z": 66.443045092804624561268904,
    "2021-11-01T00:55:00Z": 69.830178871473223212848429,
    "2021-11-01T01:00:00Z": 75.517190477307900524338388,
    "2021-11-01T01:05:00Z": 81.026761502080848269530368,
    "2021-11-01T01:10:00Z": 88.887231547107165183632002,
    "2021-11-01T01:15:00Z": 88.448713986405764872074797,
    "2021-11-01T01:20:00Z": 87.575816023048191808922289,
    "2021-11-01T01:25:00Z": 87.728664444656381667604257,
    "2021-11-01T01:30:00Z": 75.201936975037084472353900,
    "2021-11-01T01:35:00Z": 65.248495943568551519125227,
    "2021-11-01T01:40:00Z": 47.213765836935728817659740,
    "2021-11-01T01:45:00Z": 34.620016997415018965313165,
    "2021-11-01T01:50:00Z": 21.327250618902610745209004,
    "2021-11-01T01:55:00Z": 7.319217124411093071392821,
    "2021-11-01T02:00:00Z": -4.217780253032814998649976,
    "2021-11-01T02:05:00Z": -5.949413181195047076334218,
    "2021-11-01T02:10:00Z": -22.315229142790014691883591,
    "2021-11-01T02:15:00Z": -80.433002569762823909856648,
    "2021-11-01T02:20:00Z": -147.147121322294081001888455,
    "2021-11-01T02:25:00Z": -200.930010700234022244911798,
    "2021-11-01T02:30:00Z": -231.364695988360938919703171,
    "2021-11-01T02:35:00Z": -244.356375960672786013403498,
    "2021-11-01T02:40:00Z": -257.334400144556167229112832,
    "2021-11-01T02:45:00Z": -263.851833335410028099208601,
    "2021-11-01T02:50:00Z": -256.617892577983166637326841,
    "2021-11-01T02:55:00Z": -240.367735089081971285284701,
    "2021-11-01T03:00:00Z": -235.905443779098619128724225,
    "2021-11-01T03:05:00Z": -230.837774590033615369268305,
    "2021-11-01T03:10:00Z": -229.581477936606235040700068,
    "2021-11-01T03:15:00Z": -229.091997904016620993637203,
    "2021-11-01T03:20:00Z": -226.895487930859326557976700,
    "2021-11-01T03:25:00Z": -217.244142647834559389775717,
    "2021-11-01T03:30:00Z": -213.827913236372595831900009,
    "2021-11-01T03:35:00Z": -208.475278646351141809838985,
    "2021-11-01T03:40:00Z": -200.230618740820359977057967,
    "2021-11-01T03:45:00Z": -196.116077575623076661037990,
    "2021-11-01T03:50:00Z": -193.289971790062548972862361,
    "2021-11-01T03:55:00Z": -207.220629971859143530868779,
    "2021-11-01T04:00:00Z": -220.160939031991983785365330,
    "2021-11-01T04:05:00Z": -226.992680443412391576946784,
    "2021-11-01T04:10:00Z": -205.268344735056159600813203,
    "2021-11-01T04:15:00Z": -189.817445308451634153975631,
    "2021-11-01T04:20:00Z": -162.386484842497072802134506,
    "2021-11-01T04:25:00Z": -143.750986074669458377733231,
    "2021-11-01T04:30:00Z": -102.942535631788342137786556,
    "2021-11-01T04:35:00Z": -73.466484308940876797226777,
    "2021-11-01T04:40:00Z": -45.786199263303223019896730,
    "2021-11-01T04:45:00Z": -28.762771386939782333727642,
    "2021-11-01T04:50:00Z": -11.667347593356576299450122,
    "2021-11-01T04:55:00Z": 0.822427146269657217997509,
    "2021-11-01T05:00:00Z": 5.173989382201667441433715,
    "2021-11-01T05:05:00Z": 13.550011198020744768155144,
    "2021-11-01T05:10:00Z": 25.861139273685027633613186,
    "2021-11-01T05:15:00Z": 33.297366235830149221836858,
    "2021-11-01T05:20:00Z": 48.476206548213590361079486,
    "2021-11-01T05:25:00Z": 51.439956575884792503094884,
    "2021-11-01T05:30:00Z": 50.782610226329354950507576,
    "2021-11-01T05:35:00Z": 48.093434479796412652035952,
    "2021-11-01T05:40:00Z": 52.378628017643293613039255,
    "2021-11-01T05:45:00Z": 49.156159957350755407408918,
    "2021-11-01T05:50:00Z": 40.487204525983658461032911,
    "2021-11-01T05:55:00Z": 30.760960760371446796937374,
    "2021-11-01T06:00:00Z": 16.966770223092688976954405,
    "2021-11-01T06:05:00Z": 1.578538090203717021389920,
    "2021-11-01T06:10:00Z": -9.458721137446659491048237,
    "2021-11-01T06:15:00Z": -24.140804906831465482376727,
    "2021-11-01T06:20:00Z": -27.551089285616353760683929,
    "2021-11-01T06:25:00Z": -16.267966342663017183737333,
    "2021-11-01T06:30:00Z": -3.174146756672078074349119,
    "2021-11-01T06:35:00Z": 24.590801884849849468041637,
    "2021-11-01T06:40:00Z": 41.117860770545378460307423,
    "2021-11-01T06:45:00Z": 41.073615045428822788033525,
    "2021-11-01T06:50:00Z": 42.325859961130338169721685,
    "2021-11-01T06:55:00Z": 45.776184496667077136346634,
    "2021-11-01T07:00:00Z": 88.641524978963187983288157,
    "2021-11-01T07:05:00Z": 134.058626211771101746751878,
    "2021-11-01T07:10:00Z": 207.361954799164525900499787,
    "2021-11-01T07:15:00Z": 257.643913609240120027789461,
    "2021-11-01T07:20:00Z": 303.196514801074387563098585,
    "2021-11-01T07:25:00Z": 327.852296192955088480356747,
    "2021-11-01T07:30:00Z": 347.980303879551591039685998,
    "2021-11-01T07:35:00Z": 346.143497321989858389135571,
    "2021-11-01T07:40:00Z": 345.705617570943032427717025,
    "2021-11-01T07:45:00Z": 348.203511013046595763611702,
    "2021-11-01T07:50:00Z": 350.819199481782474778758681,
    "2021-11-01T07:55:00Z": 345.121300709632152357249804,
    "2021-11-01T08:00:00Z": 345.738373743307744165386727
  },
  "SignalLine": {
    "2021-11-01T00:00:00Z": 0.0,
    "2021-11-01T00:05:00Z": 2.0102564102564102564102562,
    "2021-11-01T00:10:00Z": 7.48393016290452187888085296,
    "2021-11-01T00:15:00Z": 16.132672502866381162823561368,
    "2021-11-01T00:20:00Z": 24.079625931578762297960172694,
    "2021-11-01T00:25:00Z": 31.226917809554801508702985955,
    "2021-11-01T00:30:00Z": 37.139972519541102297647321164,
    "2021-11-01T00:35:00Z": 43.256889447085456450835648531,
    "2021-11-01T00:40:00Z": 47.486891573599289973124672425,
    "2021-11-01T00:45:00Z": 50.220190541670250254426617740,
    "2021-11-01T00:50:00Z": 53.464761451897125115795074992,
    "2021-11-01T00:55:00Z": 56.737844935812344735205745794,
    "2021-11-01T01:00:00Z": 60.493714044111455893032274235,
    "2021-11-01T01:05:00Z": 64.600323535705334368331892988,
    "2021-11-01T01:10:00Z": 69.457705137985700531391914790,
    "2021-11-01T01:15:00Z": 73.255906907669713399528491232,
    "2021-11-01T01:20:00Z": 76.119888730745409081407250786,
    "2021-11-01T01:25:00Z": 78.441643873527603598646652029,
    "2021-11-01T01:30:00Z": 77.793702493829499773388101623,
    "2021-11-01T01:35:00Z": 75.284661183777310122535526698,
    "2021-11-01T01:40:00Z": 69.670482114408993861560369358,
    "2021-11-01T01:45:00Z": 62.660389091010198882310928486,
    "2021-11-01T01:50:00Z": 54.393761396588681254890543589,
    "2021-11-01T01:55:00Z": 44.978852542153163618190999071,
    "2021-11-01T02:00:00Z": 35.139525983115967894822804057,
    "2021-11-01T02:05:00Z": 26.921738150253764900591399646,
    "2021-11-01T02:10:00Z": 17.074344691645008982096401517,
    "2021-11-01T02:15:00Z": -2.427124760636557596294208386,
    "2021-11-01T02:20:00Z": -31.371124072968062277413057709,
    "2021-11-01T02:25:00Z": -65.282901398421254270912805767,
    "2021-11-01T02:30:00Z": -98.49926031640919120067087881,
    "2021-11-01T02:35:00Z": -127.67068344526191016321740265,
    "2021-11-01T02:40:00Z": -153.60342678512076157639648852,
    "2021-11-01T02:45:00Z": -175.65310809517861488095891102,
    "2021-11-01T02:50:00Z": -191.84606499173952523223249702,
    "2021-11-01T02:55:00Z": -201.55039901120801444284293782,
    "2021-11-01T03:00:00Z": -208.42140796478613538001919526,
    "2021-11-01T03:05:00Z": -212.90468128983563137786901721,
    "2021-11-01T03:10:00Z": -216.24004061918975211043522737,
    "2021-11-01T03:15:00Z": -218.81043207615512588707562250,
    "2021-11-01T03:20:00Z": -220.42744324709596602125583800,
    "2021-11-01T03:25:00Z": -219.79078312724368469495981380,
    "2021-11-01T03:30:00Z": -218.59820914906946692234785284,
    "2021-11-01T03:35:00Z": -216.57362304852580189984607927,
    "2021-11-01T03:40:00Z": -213.30502218698471351528845682,
    "2021-11-01T03:45:00Z": -209.86723326471238614443836346,
    "2021-11-01T03:50:00Z": -206.55178096978241871012316297,
    "2021-11-01T03:55:00Z": -206.68555077019776367427228618,
    "2021-11-01T04:00:00Z": -209.38062842255660769649089494,
    "2021-11-01T04:05:00Z": -212.90303882672776447258207275,
    "2021-11-01T04:10:00Z": -211.37610000839344349822829880,
    "2021-11-01T04:15:00Z": -207.06436906840508162937776524,
    "2021-11-01T04:20:00Z": -198.12879222322347986392911339,
    "2021-11-01T04:25:00Z": -187.25323099351267556668993691,
    "2021-11-01T04:30:00Z": -170.39109192116780888090926073,
    "2021-11-01T04:35:00Z": -151.00617039872242246417276398,
    "2021-11-01T04:40:00Z": -129.96217617163858257531755718,
    "2021-11-01T04:45:00Z": -109.72229521469882252699957414,
    "2021-11-01T04:50:00Z": -90.11130569043037328148968371,
    "2021-11-01T04:55:00Z": -71.924559123090367181592245168,
    "2021-11-01T05:00:00Z": -56.504849422031960256987053134,
    "2021-11-01T05:05:00Z": -42.493877298021419251958613707,
    "2021-11-01T05:10:00Z": -28.822873983680129874844253766,
    "2021-11-01T05:15:00Z": -16.398825939778074055508031413,
    "2021-11-01T05:20:00Z": -3.423819442179741172190527930,
    "2021-11-01T05:25:00Z": 7.5489357614331655628665544560,
    "2021-11-01T05:30:00Z": 16.195670654412403440394758765,
    "2021-11-01T05:35:00Z": 22.575223419489205282722997412,
    "2021-11-01T05:40:00Z": 28.535904339120022948786248930,
    "2021-11-01T05:45:00Z": 32.659955462766169440510782744,
    "2021-11-01T05:50:00Z": 34.225405275409667244615208395,
    "2021-11-01T05:55:00Z": 33.532516372402023155079641516,
    "2021-11-01T06:00:00Z": 30.219367142540156319454594213,
    "2021-11-01T06:05:00Z": 24.491201332072868459841659370,
    "2021-11-01T06:10:00Z": 17.701216838168962869663680096,
    "2021-11-01T06:15:00Z": 9.332812489168877199255598677,
    "2021-11-01T06:20:00Z": 1.9560321342118310072676931416,
    "2021-11-01T06:25:00Z": -1.6887675611631386309333120867,
    "2021-11-01T06:30:00Z": -1.9858434002649265196164734694,
    "2021-11-01T06:35:00Z": 3.3294856567580286779151486245,
    "2021-11-01T06:40:00Z": 10.887160679515498634393603500,
    "2021-11-01T06:45:00Z": 16.924451552698163465121587800,
    "2021-11-01T06:50:00Z": 22.004733234384598406041607240,
    "2021-11-01T06:55:00Z": 26.759023486841094152102612592,
    "2021-11-01T07:00:00Z": 39.135523785265512918339721474,
    "2021-11-01T07:05:00Z": 58.120144270566630684022152779,
    "2021-11-01T07:10:00Z": 87.96850637628620972731767962,
    "2021-11-01T07:15:00Z": 121.90358782287699178741203590,
    "2021-11-01T07:20:00Z": 158.16217321851647094254934572,
    "2021-11-01T07:25:00Z": 192.10019781340419445011082598,
    "2021-11-01T07:30:00Z": 223.27621902663367376802586038,
    "2021-11-01T07:35:00Z": 247.84967468570491069224780250,
    "2021-11-01T07:40:00Z": 267.42086326275253503934164700,
    "2021-11-01T07:45:00Z": 283.57739281281134718419565800,
    "2021-11-01T07:50:00Z": 297.02575414660557270310826260,
    "2021-11-01T07:55:00Z": 306.64486345921088863393657088,
    "2021-11-01T08:00:00Z": 314.46356551603025974022660210
  },
  "Histogram": {
    "2021-11-01T00:00:00Z": 0.0,
    "2021-11-01T00:05:00Z": 8.0410256410256410256410248,
    "2021-11-01T00:10:00Z": 21.89469501059244648988238704,
    "2021-11-01T00:15:00Z": 34.594969359847437135770833632,
    "2021-11-01T00:20:00Z": 31.787813714849524540546445306,
    "2021-11-01T00:25:00Z": 28.589167511904156842971253045,
    "2021-11-01T00:30:00Z": 23.652218839945203155777340836,
    "2021-11-01T00:35:00Z": 24.467667710177416612753309469,
    "2021-11-01T00:40:00Z": 16.920008506055334089156095575,
    "2021-11-01T00:45:00Z": 10.933195872283841125207781260,
    "2021-11-01T00:50:00Z": 12.978283640907499445473829008,
    "2021-11-01T00:55:00Z": 13.092333935660878477642683206,
    "2021-11-01T01:00:00Z": 15.023476433196444631306113765,
    "2021-11-01T01:05:00Z": 16.426437966375513901198475012,
    "2021-11-01T01:10:00Z": 19.429526409121464652240087210,
    "2021-11-01T01:15:00Z": 15.192807078736051472546305768,
    "2021-11-01T01:20:00Z": 11.455927292302782727515038214,
    "2021-11-01T01:25:00Z": 9.287020571128778068957604971,
    "2021-11-01T01:30:00Z": -2.591765518792415301034201623,
    "2021-11-01T01:35:00Z": -10.036165240208758603410299698,
    "2021-11-01T01:40:00Z": -22.456716277473265043900629358,
    "2021-11-01T01:45:00Z": -28.040372093595179916997763486,
    "2021-11-01T01:50:00Z": -33.066510777686070509681539589,
    "2021-11-01T01:55:00Z": -37.659635417742070546798178071,
    "2021-11-01T02:00:00Z": -39.357306236148782893472780057,
    "2021-11-01T02:05:00Z": -32.871151331448811976925617646,
    "2021-11-01T02:10:00Z": -39.389573834435023673979992517,
    "2021-11-01T02:15:00Z": -78.005877809126266313562439614,
    "2021-11-01T02:20:00Z": -115.77599724932601872447539729,
    "2021-11-01T02:25:00Z": -135.64710930181276797399899223,
    "2021-11-01T02:30:00Z": -132.86543567195174771903229219,
    "2021-11-01T02:35:00Z": -116.68569251541087585018609535,
    "2021-11-01T02:40:00Z": -103.73097335943540565271634348,
    "2021-11-01T02:45:00Z": -88.19872524023141321824968998,
    "2021-11-01T02:50:00Z": -64.77182758624364140509434398,
    "2021-11-01T02:55:00Z": -38.81733607787395684244176318,
    "2021-11-01T03:00:00Z": -27.48403581431248374870502974,
    "2021-11-01T03:05:00Z": -17.93309330019798399139928779,
    "2021-11-01T03:10:00Z": -13.34143731741648293026484063,
    "2021-11-01T03:15:00Z": -10.28156582786149510656158050,
    "2021-11-01T03:20:00Z": -6.46804468376336053672086200,
    "2021-11-01T03:25:00Z": 2.54664047940912530518409680,
    "2021-11-01T03:30:00Z": 4.77029591269687109044784384,
    "2021-11-01T03:35:00Z": 8.09834440217466009000709427,
    "2021-11-01T03:40:00Z": 13.07440344616435353823048982,
    "2021-11-01T03:45:00Z": 13.75115568908930948340037346,
    "2021-11-01T03:50:00Z": 13.26180917971986973726080197,
    "2021-11-01T03:55:00Z": -0.53507920166137985659649282,
    "2021-11-01T04:00:00Z": -10.78031060943537608887443506,
    "2021-11-01T04:05:00Z": -14.08964161668462710436471125,
    "2021-11-01T04:10:00Z": 6.10775527333728389741509580,
    "2021-11-01T04:15:00Z": 17.24692375995344747540213424,
    "2021-11-01T04:20:00Z": 35.74230738072640706179460739,
    "2021-11-01T04:25:00Z": 43.50224491884321718895670591,
    "2021-11-01T04:30:00Z": 67.44855628937946674312270473,
    "2021-11-01T04:35:00Z": 77.53968608978154566694598698,
    "2021-11-01T04:40:00Z": 84.17597690833535955542082718,
    "2021-11-01T04:45:00Z": 80.95952382775904019327193214,
    "2021-11-01T04:50:00Z": 78.44395809707379698203956171,
    "2021-11-01T04:55:00Z": 72.746986269360024399589754168,
    "2021-11-01T05:00:00Z": 61.678838804233627698420768134,
    "2021-11-01T05:05:00Z": 56.043888496042164020113757707,
    "2021-11-01T05:10:00Z": 54.684013257365157508457439766,
    "2021-11-01T05:15:00Z": 49.696192175608223277344889413,
    "2021-11-01T05:20:00Z": 51.900025990393331533270013930,
    "2021-11-01T05:25:00Z": 43.891020814451626940228329544,
    "2021-11-01T05:30:00Z": 34.586939571916951510112817235,
    "2021-11-01T05:35:00Z": 25.518211060307207369312954588,
    "2021-11-01T05:40:00Z": 23.842723678523270664253006070,
    "2021-11-01T05:45:00Z": 16.496204494584585966898135256,
    "2021-11-01T05:50:00Z": 6.261799250573991216417702605,
    "2021-11-01T05:55:00Z": -2.771555612030576358142267516,
    "2021-11-01T06:00:00Z": -13.252596919447467342500189213,
    "2021-11-01T06:05:00Z": -22.912663241869151438451739370,
    "2021-11-01T06:10:00Z": -27.159937975615622360711917096,
    "2021-11-01T06:15:00Z": -33.473617396000342681632325677,
    "2021-11-01T06:20:00Z": -29.507121419828184767951622142,
    "2021-11-01T06:25:00Z": -14.579198781499878552804020913,
    "2021-11-01T06:30:00Z": -1.1883033564071515547326455306,
    "2021-11-01T06:35:00Z": 21.261316228091820790126488376,
    "2021-11-01T06:40:00Z": 30.230700091029879825913819500,
    "2021-11-01T06:45:00Z": 24.149163492730659322911937200,
    "2021-11-01T06:50:00Z": 20.321126726745739763680077760,
    "2021-11-01T06:55:00Z": 19.017161009825982984244021408,
    "2021-11-01T07:00:00Z": 49.506001193697675064948435526,
    "2021-11-01T07:05:00Z": 75.938481941204471062729725221,
    "2021-11-01T07:10:00Z": 119.39344842287831617318210738,
    "2021-11-01T07:15:00Z": 135.74032578636312824037742510,
    "2021-11-01T07:20:00Z": 145.03434158255791662054923928,
    "2021-11-01T07:25:00Z": 135.75209837955089403024592102,
    "2021-11-01T07:30:00Z": 124.70408485291791727166013762,
    "2021-11-01T07:35:00Z": 98.29382263628494769688776850,
    "2021-11-01T07:40:00Z": 78.28475430819049738837537800,
    "2021-11-01T07:45:00Z": 64.62611820023524857941604400,
    "2021-11-01T07:50:00Z": 53.79344533517690207565041840,
    "2021-11-01T07:55:00Z": 38.47643725042126372331323312,
    "2021-11-01T08:00:00Z": 31.27480822727748442516012490
  }
}
